### ISSUE TITLE
remove some interference between session in the client

### ISSUE THREAD

**Issue by smithdh** [xrootd/xrootd#1942]
There's an ongoing investigation into some occasional failures seen on an EOS instance at cern; While that's still going on I've seen some problematic behavior in XrdCl that is probably contributing. But I think there's enough to open an issue for the specifics seen in the client so far. Illustrating with this snippet:

```
XrdCl::File *f = new XrdCl::File();
f->Open("root://me@pmpe01:1194//f1", XrdCl::OpenFlags::Read);
f->Read(0,5,buf,bytesRead,0); // gets 5 bytes from file 'f1' ok

sleep(20); // during this time the Xrootd server on port 1194 is restarted

XrdCl::File *f2 = new XrdCl::File();
f2->Open("root://me@pmpe01:1194//pww1", XrdCl::OpenFlags::Read);
f2->Read(0,5,buf,bytesRead,0); // gets bytes from file 'pww1' ok
f2->Close();
delete f2;

sleep(330);
// during this time the default TTL (300s for dataserver) expires and connection is closed

f2 = new XrdCl::File();
f2->Open("root://me@pmpe01:1194//pww1", XrdCl::OpenFlags::Read);
f2->Read(0,5,buf,bytesRead,0); // reads bytes from 'pww1' ok

f->Read(0,5,buf,bytesRead,0); // reports no error but bytes are from 'pww1' not 'f1'
```

(In the case of the EOS problem this is probably only a contributing factor, there may be other issues. For instance there is no server-restart, unlike in the above - possibly we have a stream-timeout instead).

**Comment by smithdh** [comment]
A variation can give unexpected errors rather than unexpected data:

```
XrdCl::File *f = new XrdCl::File();
f->Open("root://me@pmpe01:1194//f1", XrdCl::OpenFlags::Read);
f->Read(0,5,buf,bytesRead,0); // gets 5 bytes from file 'f1' ok

sleep(20); // during this time the Xrootd server on port 1194 is restarted

XrdCl::File *f2 = new XrdCl::File();
f2->Open("root://me@pmpe01:1194//pww1", XrdCl::OpenFlags::Read);
f2->Read(0,5,buf,bytesRead,0); // gets bytes from file 'pww1' ok
f2->Close();
delete f2;

sleep(330);
// during this time the default TTL (300s for dataserver) expires and connection is closed

f2 = new XrdCl::File();
f2->Open("root://me@pmpe01:1194//pww1", XrdCl::OpenFlags::Read);
f2->Read(0,5,buf,bytesRead,0); // reads bytes from 'pww1' ok
f->Close();
delete f;

f2->Read(0,5,buf,bytesRead,0); // gives an error "read does not refer to an open file" or EBADF.
```

**Comment by simonmichal** [comment]
Hi @smithdh, thanks for reporting this issue, there's definitely something off, the TTL should not kick in if there are open files on the connection:
https://github.com/xrootd/xrootd/blob/31d6122520f5f036e046b1cdd5567e410548b02b/src/XrdCl/XrdClXRootDTransport.cc#L772

is this run against vanilla xrootd server or eos?

**Comment by smithdh** [comment]
Hi @simonmichal . Yes, I think the example should show the problem with a vanilla xrootd server. I've tried to make a fix possible fix (#1945) - taking what I felt was a defensive approach of keeping the Stream sessionId unique also between (re)-created Stream objects.

**Comment by amadio** [comment]
Closing, as #1945 has been merged.
